### ISSUE TITLE
DCBackend: deltacnn, delta_cudnn

### ISSUE THREAD

**Issue by PHaiJun** [dabeschte/DeltaCNN#3]
what's the difference of two choices of DCBackend, deltacnn and delta_cudnn?

**Comment by dabeschte** [comment]
`deltacnn` is using our sparse cuda kernels that are able to accelerate sparse inference. 
`delta_cudnn` simulates the same behaviour, but is actually performing a dense inference using PyTorch's standard cuDNN backend and is much slower than standard dense inference. However, this mode is differentiable (it supports backpropagation).

We used it internally for some experiments where we tried to train the truncation thresholds instead of auto tuning them. This can be done using a soft truncation (scaling down values below the threshold) instead of hard truncation (setting them to zero) of updates. However, it turned out that threshold training is much trickier to implement and the resulting sparsity/accuracy ratio did not improve over auto tuning - which is why we did not use them in the end.

**Comment by PHaiJun** [comment]
Does the standard dense inference you said mean another mode of DCBackend(cudnn=0), that is to say,  delta_cudnn is much slower than cudnn?

**Comment by dabeschte** [comment]
with standard dense inference I meant the `cudnn` backend. `delta_cudnn` is performing a dense `cudnn` convolution + overhead (getting delta features, performing a soft truncation, adding delta onto previous values, etc.).
I don't know how much slower it is exactly, but it definitely was significant, because this mode is unoptimized.
